### ISSUE TITLE
Feature Request: Async activation offloading to CPU with prefetch

### ISSUE THREAD

**Issue by niyunsheng** [pytorch/torchtitan#2379]
### Motivation

torchtitan currently supports **selective activation checkpointing** (selective AC), which saves
the outputs of compute-intensive ops like `mm` and flash attention to avoid recomputing them in
backward. This effectively eliminates the major FLOP overhead of checkpointing.

However, selective AC still operates within **block-wise checkpointing** — the entire transformer
block is checkpointed, and all ops not in the save list are recomputed. This means memory-bound
ops like layernorm, activation functions, dropout, and other elementwise ops are still redundantly
re-executed in backward. While these ops contribute minimal FLOPs, they are **bandwidth-bound** and
consume real GPU execution time, reducing overall training throughput.

The most performant approach is to **eliminate block-wise checkpointing entirely** and instead use
**async activation offloading**: offload all activations to CPU during forward and prefetch them
back before backward. This avoids all redundant recomputation — including the memory-bound ops
that selective AC cannot eliminate — while the D2H/H2D transfers can be fully overlapped with
computation via separate CUDA streams, achieving near-zero overhead.

### Current state

- torchtitan has activation checkpointing (`torchtitan/distributed/activation_checkpoint.py`)
  with modes: `full`, `selective`, `memory_budget`.
- **No activation offloading support exists.**
- I have filed a feature request in PyTorch core (
  https://github.com/pytorch/pytorch/issues/174960) proposing to
  integrate activation offloading into FSDP's `CPUOffloadPolicy`. However, PyTorch core
  features have long design and review cycles, and this may take a while.
- In the meantime, torchtitan can implement this at the application level using
  `torch.autograd.graph.saved_tensors_hooks`, without waiting for PyTorch core support.

### Proposed usage

```python
for layer in model.layers:
    with activation_offload():
        x = layer(x)
```

Activation offloading is orthogonal to checkpointing — they can be used independently
or composed together.

### Implementation approach

I suggest following [Megatron-Kwai's `OffloadManager`](https://github.com/kwai/Megatron-Kwai/blob/sc25slimpipe/megatron/core/pipeline_parallel/offload.py)
(BSD 3-Clause licensed, compatible with torchtitan's BSD license). 

Key design points:
1. **`saved_tensors_hooks`**: Use `pack_hook` to collect activation tensors during forward,
then batch offload to CPU; use `unpack_hook` to return prefetched tensors in backward.
2. **Pre-allocated pinned memory pool**: Avoid dynamic pinned memory allocation which triggers
CUDA implicit synchronization and prevents stream overlap
(known issue: https://github.com/meta-pytorch/torchtune/issues/2076).
3. **Ping-pong GPU buffers**: Two alternating GPU buffers for prefetch — one serves backward
computation while the other receives the next layer's activations from CPU.
4. **Separate D2H/H2D CUDA streams**: Overlap PCIe transfers with computation on the default
stream.
5. **Default prefetch = 1 layer ahead**: In backward, when computing layer N, prefetch
layer N-1's activations from CPU.

### Willing to contribute

I have experience implementing async activation offloading based on the Megatron-Kwai approach.

If there is interest in this feature, I can prepare a PR with proper attribution to the original
Megatron-Kwai implementation per its BSD 3-Clause license.

### Configuration

A possible config option under [training], alongside the existing enable_cpu_offload
for parameter/gradient/optimizer state offloading:

```yaml
[training]
enable_activation_offload = true
```

### Related
* https://github.com/pytorch/pytorch/issues/174960 - my proposal to
add activation offloading natively in FSDP's CPUOffloadPolicy (long-term solution)
* https://github.com/pytorch/pytorch/issues/158657 - related proposal
to extend `torch.utils.checkpoint` with offloading
* [torchtune `OffloadActivations`](https://github.com/meta-pytorch/torchtune/blob/main/torchtune/training/_activation_offloading.py) — torchtune's implementation (has stream sync issues)
* [Megatron-Kwai `OffloadManager`](https://github.com/kwai/Megatron-Kwai/blob/sc25slimpipe/megatron/core/pipeline_parallel/offload.py) — reference implementation with ping-pong buffers

cc @fegin  @soulitzer @wanchaol

**Comment by tianyu-l** [comment]
cc @soulitzer @H-Huang

**Comment by dean-mccoppin** [comment]
Wanted to try this out. Couldn't test multi-GPU FSDP integration due to hardware limitations, so I'd verify just in case if someone's capable. Thanks!
